pkg/libdevfile/component_kubernetes.go

```diff
@@ -27,3 +28,70 @@ func (e *kubernetesComponent) CheckValidity() error {
 func (e *kubernetesComponent) Apply(handler Handler) error {
 	return handler.ApplyKubernetes(e.component)
 }
+
+// GetK8sAndOcComponentsToPush returns the list of Kubernetes and OpenShift components to push,
+// The list returned is governed by the DeployByDefault field in each component.
+// All components with DeployByDefault set to true are included, along with those with no DeployByDefault set and not-referenced.
+// It takes an additional allowApply boolean, which set to true, will append the components referenced from apply commands to the list.
+func GetK8sAndOcComponentsToPush(devfileObj parser.DevfileObj, allowApply bool) ([]v1alpha2.Component, error) {
+	var allComponents []v1alpha2.Component
+
+	k8sComponents, err := devfileObj.Data.GetComponents(parsercommon.DevfileOptions{
+		ComponentOptions: parsercommon.ComponentOptions{ComponentType: v1alpha2.KubernetesComponentType},
+	})
+	if err != nil {
+		return nil, err
+	}
+	allComponents = append(allComponents, k8sComponents...)
+
+	ocComponents, err := devfileObj.Data.GetComponents(parsercommon.DevfileOptions{
+		ComponentOptions: parsercommon.ComponentOptions{ComponentType: v1alpha2.OpenshiftComponentType},
+	})
+	if err != nil {
+		return nil, err
+	}
+	allComponents = append(allComponents, ocComponents...)
+
+	allApplyCommands, err := devfileObj.Data.GetCommands(parsercommon.DevfileOptions{
+		CommandOptions: parsercommon.CommandOptions{CommandType: v1alpha2.ApplyCommandType},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	m := make(map[string]v1alpha2.Component)
+	for _, comp := range allComponents {
+		if comp.Kubernetes == nil && comp.Openshift == nil {
+			continue
+		}
```

Comment:
Same comment as above, I don't think this condition will ever be hit since  we are filtering the components based on their dedicated types.

Reply:
Done in [`9544a55` (#6720)](https://github.com/redhat-developer/odo/pull/6720/commits/9544a55c4c0165d95fa9a90122597fb12e2f75ed)